pkg/operator/controllers/base/aro_controller.go

```diff
@@ -0,0 +1,105 @@
+package base
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+)
+
+type AROController struct {
+	Log    *logrus.Entry
+	Client client.Client
+	Name   string
+}
+
+func (c *AROController) SetConditions(ctx context.Context, cnds ...*operatorv1.OperatorCondition) {
+	cluster, err := c.getCluster(ctx)
+	if err != nil {
+		c.Log.Warn("Failed to retrieve ARO cluster resource")
+		return
+	}
+
+	newConditions := cluster.Status.DeepCopy().Conditions
+	for _, cnd := range cnds {
+		v1helpers.SetOperatorCondition(&newConditions, *cnd)
+	}
+
+	if equality.Semantic.DeepEqual(cluster.Status.Conditions, newConditions) {
+		return
+	}
+
+	cluster.Status.Conditions = newConditions
+	if err := c.Client.Status().Update(ctx, cluster); err != nil {
+		c.Log.Error("error updating controller conditions", err)
+	}
+}
+
+func (c *AROController) SetProgressing(ctx context.Context, message string) {
+	cnd := c.defaultProgressing()
+	cnd.Status = operatorv1.ConditionTrue
+	cnd.Message = message
+
+	c.SetConditions(ctx, cnd)
+}
+
+func (c *AROController) ClearProgressing(ctx context.Context) {
+	c.SetConditions(ctx, c.defaultProgressing())
+}
+
+func (c *AROController) SetDegraded(ctx context.Context, err error) {
+	cnd := c.defaultDegraded()
+	cnd.Status = operatorv1.ConditionTrue
+	cnd.Message = err.Error()
+
+	c.SetConditions(ctx, cnd)
+}
+
+func (c *AROController) ClearDegraded(ctx context.Context) {
+	c.SetConditions(ctx, c.defaultDegraded())
+}
+
+func (c *AROController) ClearConditions(ctx context.Context) {
+	c.SetConditions(ctx, c.defaultAvailable(), c.defaultProgressing(), c.defaultDegraded())
+}
+
+func (c *AROController) getCluster(ctx context.Context) (*arov1alpha1.Cluster, error) {
```

Comment:
Maybe make this method public (`GetCluster`) since it looks like quite a lot of our controllers fetch this object.  This a good example of controller commonalities I mentioned earlier.

Reply:
Made the change, and adapted the machineset and node controllers to use this method instead of doing the retrieval themselves. There's probably more opportunities for further common functionality refactors along this line (these two controllers both contain the same logic to bail early if their respective feature flags in the cluster resource are not enabled), but I'll leave that for future consideration.